clang: Always pass POSIX paths to consumers

amyspark opened the pull request:
:wave:

This is a PR that fixes #12191 (at least to the extent that my local project stash could check).

It requires #12144 to fix argument detection.

DFOVIT commented:
Feel free to ignore as this is only tangentially related to your problem but i'm pretty sure that [this line](https://github.com/mesonbuild/meson/blob/06ca5fd7d6ed6e4ab3104578fdb44f2bbb5ab2a9/mesonbuild/backend/backends.py#L1975C90-L1975C90) needs the same Path().as_posix() treatment because meson devenv will generate a vscode env file with paths like c:\folder and you will get surprises when your folder name start with a 'n'. I tested with the Path().as_posix() and it generates a correct vscode env file for windows.

eli-schwartz commented on mesonbuild/compilers/mixins/clang.py:
```diff
@@ -121,6 +121,13 @@ def openmp_flags(self) -> T.List[str]:
             # Shouldn't work, but it'll be checked explicitly in the OpenMP dependency.
             return []
 
+    def gen_vs_module_defs_args(self, defsfile: str) -> T.List[str]:
```
The commit message doesn't really explain why this diff hunk should be necessary, which feels like a sign that either the commit message is wrong, the change is wrong, or this should be two commits.

amyspark replied:
Changing how the paths are forwarded will reveal that module definition files are passed with the wrong flags when using clang+link. I found this during the testing of this MR, but you are right -- this belongs to https://github.com/mesonbuild/meson/pull/12144.

xclaesse commented:
I'm a bit worried that scattering millions of `Path().as_posix()` and `replace('\\', '/')` in random places of Meson is going to be slow and inconsistent. Most of the time it is to workaround broken tools that should instead get fixed.

Alternatively, we could make a project wide search-and-replace to ban usage of `os.path.join()` and `str(Path())`, and ensure all our paths are always posix from the beginning, instead of fixing them in random places.

amyspark commented:
> Most of the time it is to workaround broken tools that should instead get fixed.

I could agree -- in this case, Microsoft Clang not accepting Windows-style paths.

But, if you check the changes that are necessary in the unit tests, this assertion also applies to *Meson itself* 😃 so the proper fix would be to use one or the other consistently. But there are 300+ places where this is used, so I don't quite know which ones are for internal consumption only and which ones do are passed externally.

amyspark commented:
@eli-schwartz I just found https://github.com/mesonbuild/meson/pull/12564 which solves a lot of the issues I was facing. However, there are a couple spots that still needed fixing when using Microsoft clang. I've rebased and fixed them here, and rephrased the commit to clarify that.

nirbheek reviewed (approved):
LGTM, since all these are places where we were using os.path.join and now are using as_posix().

For future reference: replacing `os.path.join` good, and `str.replace('\\', '/')` BAD BAD BAD.
